### ISSUE TITLE
Layout and behaviour of Recurring Transaction dialog is not consistent with Single Transaction dialog

### ISSUE THREAD

**Issue by tactilis** [moneymanagerex/moneymanagerex#4772]
MMEX version:
 - [x] 1.5.17 Beta1

Operating System:
 - [x] Windows
 - [ ] Mac OSX
 - [ ] Linux 

# Description of the bug

The layout and behaviour of some controls on the Recurring Transaction (RT) dialog is not consistent with the same controls on the Single Transaction (ST) dialog.

# Reproduction

Is the bug reproducible? 
 - [x] Always 
 - [ ] Randomly 
 - [ ] Happened only once

Reproduction steps:

Compare the field sizes on the screenshots below and in the video where I have overlaid the ST dialog on top of the RT dialog.

![image](https://user-images.githubusercontent.com/4060092/177017108-38bab697-0876-4239-b27f-06352ed7cad0.png)
 
https://user-images.githubusercontent.com/4060092/177017042-0cc574e9-69bc-4c33-9526-e5e8adbee6b4.mp4
 
 
And observe in the video below what happens to the layout of RT dialog if you:
1. Click in the empty Category field
2. Click on the 'Use split Categories' button
3. Cancel the Split Transaction dialog

https://user-images.githubusercontent.com/4060092/177028889-6b2caee3-3981-459a-85a2-92233200ba35.mp4

Expected result:

1. The layout and behaviour of the RT and ST dialogs should be consistent. They should use the same code (not just a copy) where possible to ensure that improvements to the ST dialog appear in the RT dialog too.

2. The Category and Notes fields should be the same width and in the same position on the RT dialog as on the ST dialog.

3. The Color button on the RT dialog should have a label as on the ST dialog.

4. The behaviour of Payee field on the RT dialog should be identical to that on the ST dialog; at present pressing Enter in the RT Payee field does **not** display the Organize Payees dialog.

5. The behaviour of Category field on the RT dialog should be identical that on the ST dialog; at present pressing Enter in the RT Category field does **not** display the Organize Categories dialog.

6. The Category field should not increase in width and the 'Use split Categories' button should not change position (as shown in the 2nd video above)

**Comment by SJK55** [comment]
Additional information for Recurring Transactions. Seems the application identifies new Payee but can't act on it even though it says would you like to add. Fault. Images of both normal and recurring transaction acceptance of new payee.
![Payee Good add](https://user-images.githubusercontent.com/14268231/177029800-d50b06d7-9fae-4972-a7ee-b6cc99ba7a69.jpg)
![Payee Not accepted Recurring Transaction](https://user-images.githubusercontent.com/14268231/177029813-bc1804a5-294c-4526-a214-b876426fbd9a.jpg)

**Comment by whalley** [comment]
Probably worth throwing into the mix is the Share Transaction dialog, which although slightly different is now inconsistent with the other dialogs.

![CleanShot 2022-07-03 at 09 25 33](https://user-images.githubusercontent.com/487345/177031574-014fd12b-87a8-4aad-a1da-1f0b31a2ec9f.png)

**Comment by vomikan** [comment]
Guys, I don't want to upset you, but there is another dialogs related to assets and shares. Few people use it, but for the most part it also repeats the transaction dialog.
I made all the changes to the transaction dialog first. There were not enough forces to change the dialog of recurring transactions, so it always lagged behind in development. Later, dialogs for shares and assets appeared. The programmer's hand did not reach them after.
Unifying all dialogs is a difficult task. Moreover, I don't like the (complicated) realization of dilogs for shares  and assets at all.

**Comment by tactilis** [comment]
Agreed, @whalley.

How possible it is to abstract the behaviour (and layout) of various parts of these dialogs to one or more classes, as you have done with the date picker, so that the behaviour is guaranteed to be the same across dialogs?

When you look at the Recurring Transaction 'Transaction details' panel it's obvious that it should be an instance of the Single Transaction 'Transaction details' panel. Can the code be refactored in this way?

Additionally, although I didn't show it above, the same consideration applies to the 'Custom fields' panel of the Single, Recurring, and Share transaction dialogs. Or is that already handled by a single class?

The purpose of this Issue/Bug report is not just to get the RT and ST dialogs to look and behave consistently now... but to ensure that they will always be consistent in the future in the face of updates to functionality.

**Comment by tactilis** [comment]
@vomikan 
Unifying all dialogs is a difficult task.

Yes, I appreciate that fact. But Recurring Transaction and Single Transaction clearly should meant to be the same because RTs create STs.


> Moreover, I don't like the (complicated) realization of dilogs for shares and assets at all.

Agreed. That needs separate work to decide how that whole area of functionality is supposed to work!

**Comment by whalley** [comment]
@vomikan The asset and share transaction dialogs already use a shared panel `UserTransaactionPanel`, I'll work on getting this up-to-date then we can look at how we can merge this into the other transaction dialogs. That way we can migrate in stages without disrupting the major dialogs in the interim.

**Comment by tactilis** [comment]
@whalley 
> The asset and share transaction dialogs already use a shared panel UserTransaactionPanel, I'll work on getting this up-to-date then we can look at how we can merge this into the other transaction dialogs.

That's good but isn't there a wider issue regarding Assets (and maybe Shares too) in that nobody really understands fully how they work (or why they were implemented they way they were). We've often had comments on the forum on on github saying this.

Rather than fix up Assets (and Shares?) transaction panels now, would it be better to stand back and understand how they _should_ operate and then implement this?

**Comment by tactilis** [comment]
@tactilis
> Rather than fix up Assets (and Shares?) transaction panels now...

Ah, I've just seen your UPDATE above @whalley

**Comment by whalley** [comment]
> The layout and behaviour of the RT and ST dialogs should be consistent. They should use the same code (not just a copy) where possible to ensure that improvements to the ST dialog appear in the RT dialog too.

As noted by @vomikan unification of the dialogs is a large task and could introduce more issues and we want a bit more stability in this area so we should cover this later - https://github.com/moneymanagerex/moneymanagerex/issues/4775 raised to cover this. We will for now fix any notable discrepancies.

> The Category and Notes fields should be the same width and in the same position on the RT dialog as on the ST dialog.

Fixed.

> The Color button on the RT dialog should have a label as on the ST dialog.

Fixed.

> The behaviour of Payee field on the RT dialog should be identical to that on the ST dialog; at present pressing Enter in the RT Payee field does not display the Organize Payees dialog.

Fixed.

> The behaviour of Category field on the RT dialog should be identical that on the ST dialog; at present pressing Enter in the RT Category field does not display the Organize Categories dialog.

Fixed.

> The Category field should not increase in width and the 'Use split Categories' button should not change position (as shown in the 2nd video above)

Fixed.

> Additional information for Recurring Transactions. Seems the application identifies new Payee but can't act on it even though it says would you like to add. Fault. Images of both normal and recurring transaction acceptance of new payee.

Fixed.

--

Note, in aligning the dialogs, I have not added the 'Auto Number' button as this is not appropriate for recurring transactions... Maybe what we need is a checkbox to specify the number should be auto-generated when each transaction is posted --- but that's another feature!

**Comment by tactilis** [comment]
>  I have not added the 'Auto Number' button as this is not appropriate for recurring transactions.

Correct. Which is why I did not include that as a 'fault'  :+1: 

Well done on the quick fixes @whalley   :1st_place_medal: 

I'll retest when I have the chance.


Until #4775 is addressed, we must now remember to keep RT synchronised with **_all_** relevant changes to the ST dialog.

**Comment by tactilis** [comment]
Retested. All fixed now.

The only (very minor) thing that is different between the previous and the current versions of the dialog is the vertical positioning of the day name in relation to the date field. This is due to the use of the new DatePicker control.

The day name is now not on the centre line of the date control; it is raised upwards a little.

![image](https://user-images.githubusercontent.com/4060092/177056412-0be0ae66-3536-4feb-91d8-5ef75c41f228.png)

Comparing old with new:

![image](https://user-images.githubusercontent.com/4060092/177056740-663e6a5f-77e9-4cf7-a9be-8be5c7fbd3bc.png)

![image](https://user-images.githubusercontent.com/4060092/177056754-69ad8655-bc3e-4efd-9cf3-f11dfe5b76ae.png)

Trivial really, but it looks better if the label, date field, spinbox, _and_ day name are all aligned.

Any change will, of course, fix it for all dialogs.

**Comment by whalley** [comment]
> The day name is now not on the centre line of the date control; it is raised upwards a little.

Yes, that came about from this recent fix: https://github.com/whalley/moneymanagerex/commit/9cef9a9c2e99f0bc375b27303701126dd6ad5eee 

In its original position the lower part of the name was being clipped on MacOS. Have found a better fix that will allow the wording to be moved back without clipping.

**Comment by whalley** [comment]
https://github.com/whalley/moneymanagerex/commit/5fd810717663c8f2b5ab6ccbba9250ece68fb5a2
